### DISCUSSION TITLE
Different versions of packages

### DISCUSSION THREAD

**Discussion by LourensVeen** [amusecode/amuse#1032] [General]
I'm looking at building different versions of packages again, for example with or without MPI, or with and without GPU support (and therefore say a CUDA dependency). A community code has an `interface.py`, possibly other Python code (?), and one or more worker binaries. Depending on what hardware you have you may be able to use one or more workers, but is `interface.py` always the same, regardless of which worker is used and regardless of global features like MPI?

I'm thinking that the best solution is to have `amuse-code` with `interface.py` and `cpu_worker`, and then have a separate `amuse-code-gpu` package with only `gpu_worker`, which would depend on `amuse-code` (and on CUDA or OpenCL or something). So in general there would be a base package with the interface and a basic works-anywhere worker if available, and then the other packages would be add-ons that provide additional workers. Does that make sense, or would this break something?

For MPI support, I guess we need a version of the framework with MPI support and workers that are built with MPI support, which would then depend on `amuse-framework-mpi` instead of `amuse-framework`, but does this change anything in `interface.py`? Is there a reason to have a non-MPI version at all in a desktop environment? I suppose we need it mainly for HPC machines that don't do MPI_Spawn?

**Comment by rieder** [comment]
I think this is largely the way to go. `amuse-code` and `amuse-code-gpu` as an additional package makes sense.
I'm not sure yet about the framework, since I don't know of any desktop systems where MPI is really not available and having multiple frameworks available could be confusing.

**Reply by rieder** [reply]
Other python code for packages could be e.g. the test module. And also `__init__.py`.
Some packages would additionally have a `data` directory.

**Reply by rieder** [reply]
As for `interface.py`, this is indeed always the same file.

**Comment by LourensVeen** [comment]
Okay. I seem to have managed to convince setuptools to build a Python wheel without any Python code in it :smile:. So we can package up additional workers separately, and put everything Python into a base package.

Setuptools is quite particular about how it wants the code to be organised, and I've been thinking about reorganising the repository, but that can be postponed and I'd like to keep the build system/packaging PR as small as possible, it's going to be a big one anyway. So maybe for the future.

**Comment by LourensVeen** [comment]
Well, it looks like I spoke too soon. Initially it seemed like setuptools was doing just what I wanted, but then I started running into all sorts of intermittent problems, and in the end I looked around the Internet a bit and discovered that Hatchling works better. It looks like we'll be using that instead, at least by default, community codes will be able to use something different if they want.

My prototype now has the following mock-up community code (I don't know offhand if ph4 has a GPU and a non-GPU version, but in the mock-up it does):

```
amuse/community/ph4
├── cpu_worker.c                        # Existing workers and Python files work in the same way
├── gpu_worker.c
├── __init__.py
├── interface.py
├── Makefile                            # This Makefile is different from what we had
├── src                                 # Still the source and the original build system
│   ├── Makefile
│   └── ph4.c
├── packages                            # Package definitions now per community code
│   ├── amuse-ph4.amuse_deps            # This declares dependencies for the top-level build system
│   ├── amuse-ph4
│   │   ├── ph4 -> ../..                # Needed for the packager to find the files
│   │   └── pyproject.toml              # Package configuration
│   └── amuse-ph4-gpu.amuse_deps
│   ├── amuse-ph4-gpu
│   │   ├── ph4 -> ../..
│   │   └── pyproject.toml
└── support                             # Local autoconf-based configuration
    ├── aclocal.m4
    ├── config.mk.in
    ├── configure
    ├── configure.ac                    # v Symlink for shared m4 macros and a helper script
    └── shared -> ../../../../../support/shared
```
with the `amuse-ph4` package containing `__init__.py`, `interface.py` and `cpu_worker`, and the `amuse-ph4-gpu` package containing only `gpu_worker`. So then if you don't have CUDA you can just use `amuse-ph4` and if you do then you install both and select the GPU worker in your script.

At the top level, `make configure` produces a bunch of output from the configure script, and then
```
*** Configuration complete ***

Detected features: forsockets gcc g++ gfortran java mpi openmp blas lapack gsl gmp mpfr fftw hdf5

** Enabled packages **

amuse-bhtree 
amuse-ph4

** Disabled packages **

amuse-ph4-gpu (missing features: cuda)

*** Next steps ***

* Enable more packages *

Some packages are disabled due to missing features. You can enable more packages by
installing additional software. Some software does require specific hardware, for
example CUDA requires an nVidia GPU to work.
 
Installing libraries using APT

We seem to be running on Ubuntu or a similar Linux distribution that uses APT.
Here's how to install the required packages for all codes system-wide using apt:


    sudo apt install gcc g++ gfortran python3 pkg-config cmake libopenmpi-dev openmpi-bin libgsl-dev libfftw3-3 libfftw3-dev libgmp3-dev libmpfr6 libmpfr-dev libhdf5-serial-dev hdf5-tools libnetcdf-dev liblapack-dev libblas-dev

 
Installing libraries using Conda

If you don't have administrator access to this machine, then using the system package
manager will not work. Instead, you can install the dependencies into the local conda
environment:


    conda install -c conda-forge gcc gxx gfortran python pkgconfig cmake openmpi openmpi-mpicc openmpi-mpicxx openmpi-mpifort gsl fftw gmp mpfr hdf5 netcdf4 libopenblas liblapack

 
Installing CUDA

If you have an nVidia GPU and a code that can use CUDA, then compiling with CUDA support
will greatly increase performance. Installing CUDA is a bit more complicated than the
other dependencies however. CUDA consists of two parts, the driver and the toolkit. The
driver must be installed on the system, as part of the nVidia GPU drivers. The toolkit
can be installed separately via an installer, the system package manager, or conda.

Here are a few useful links:

Installing nVidia drivers on Ubuntu: https://ubuntu.com/server/docs/nvidia-drivers-installation
NVIDIA CUDA installation guide for Linux: https://docs.nvidia.com/cuda/cuda-installation-guide-linux/index.html
CUDA on Windows Subsystem for Linux: https://docs.nvidia.com/cuda/wsl-user-guide/index.html#getting-started-with-cuda-on-wsl-2
 
To install the CUDA toolkit using conda, you can use

    conda install -c nvidia cuda-toolkit

but note that you will still need to install the drivers separately if they are not
available already.


* Install AMUSE *

To install the AMUSE framework and all of the enabled packages into the active
conda environment pip-test, type

    make install-packages


To install a specific package, you can use

    make install-amuse-bhtree

or whichever package you want to install, as long as it's enabled.

To do a development install of a specific package, you can use

    make develop-amuse-bhtree

or whichever package you want to work on.
```
Packages are declared enabled or disabled based on the list of features detected by autoconf and the `amuse-*.amuse_deps` files describing which features are needed. If all the dependencies are available then we expect the package to build and a failure is a real failure, whereas if a dependency is missing we won't even try building the package (in fact, the top level `install-amuse-ph4-gpu` make target doesn't even exist if you don't have CUDA). You can still move into the code directory and try to build it there, although of course it'll fail.

The top-level build system acts more like an installer than a build system, it'll help you install dependencies and install all packages or individual packages as desired.

The help text that's printed will change depending on the platform you're on and whether you have a Conda environment activated. So if you're on a Mac, it'll detect either Homebrew or MacPorts and tell you how to install things using that instead of via apt, and if you don't have a Conda environment active it'll tell you how to make one.
